File: content/snippets/python/interview_lcci/01/05_one_away.py
def one_away(a: str, b: str) -> bool:
    if abs(len(a) - len(b)) > 1:
        return False

    s1, s2 = (a, b) if len(a) <= len(b) else (b, a)
    i = j = 0
    found = False

    while i < len(s1) and j < len(s2):
        if s1[i] == s2[j]:
            i += 1
            j += 1
            continue

        if found:
            return False
        found = True

        if len(s1) == len(s2):
            i += 1
        j += 1

    return True


def main() -> None:
    print(f"pale vs ple -> {one_away('pale', 'ple')}")
    print(f"pales vs pale -> {one_away('pales', 'pale')}")
    print(f"pale vs bake -> {one_away('pale', 'bake')}")


if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci/01/05_one_away.ts
function oneAway(a: string, b: string): boolean {
  if (Math.abs(a.length - b.length) > 1) {
    return false;
  }

  let s1 = a;
  let s2 = b;
  if (s1.length > s2.length) {
    [s1, s2] = [s2, s1];
  }

  let i = 0;
  let j = 0;
  let found = false;

  while (i < s1.length && j < s2.length) {
    if (s1[i] === s2[j]) {
      i += 1;
      j += 1;
      continue;
    }

    if (found) {
      return false;
    }
    found = true;

    if (s1.length === s2.length) {
      i += 1;
    }
    j += 1;
  }

  return true;
}

function main(): void {
  console.log(`pale vs ple -> ${oneAway("pale", "ple")}`);
  console.log(`pales vs pale -> ${oneAway("pales", "pale")}`);
  console.log(`pale vs bake -> ${oneAway("pale", "bake")}`);
}

main();
